### DISCUSSION TITLE
Problem with TelegramDesktop WebApp in i3wm

### DISCUSSION THREAD

**Discussion by JustCryingCat** [i3/i3#5091] [Q&A]
## I'm submitting a…
<pre>
[x] Bug
[ ] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior
I get this result. Just a white window and a message in the console "Could not determine the accessibility bus address" (Compton has nothing to do with)
![image](https://user-images.githubusercontent.com/90184510/185618285-148e5bc8-ca7c-4f1a-9a17-fed7a9fa1260.png)

## Expected Behavior
![image](https://user-images.githubusercontent.com/90184510/185618392-d49316fb-ebfd-4b97-a47f-6c8b801e64b4.png)

## Reproduction Instructions
1. I downloaded telegram-desktop from snap
2. Running /tsetup.4.1.0/Telegram/Telegram
3. Open Web App durger king bot

## Environment
Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 
Binary i3 version:  4.20.1 (2021-11-03) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.20.1 (2021-11-03) (pid 24327)
Loaded i3 config:
  /home/egorka/.config/i3/config (main)
</pre>

<details><summary>Config file</summary><pre>
[https://pastebin.com/MhSDh7Dv](https://pastebin.com/MhSDh7Dv)
</pre>
</details>

## Logs
<pre>
i3 Logfile URL: [https://logs.i3wm.org/logs/6279115220975616.bz2](https://logs.i3wm.org/logs/6279115220975616.bz2)
</pre>

Telegram Logs In ~/.local/share/TelegramDesktop/DebugLog/log_*_*.txt:
<pre>
[21:46:55.257 00-0004272] BotWebView Error: Could not parse """".
[21:46:55.257 00-0004273] BotWebView Error: Could not parse """".
[21:46:55.290 00-0004274] BotWebView Error: Could not parse """".
</pre>

In terminal:
<pre>
Could not determine the accessibility bus address
</pre>

In journalctl:
<pre>
авг 19 18:03:14 egorka-P5K WebKitWebProces[32702]: Locale not supported by C library.
                                                              Using the fallback 'C' locale.
авг 19 18:03:14 egorka-P5K WebKitNetworkPr[32699]: GDBus.Error:org.freedesktop.portal.Error.NotAllowed: This call is not available inside the sandbox
авг 19 18:03:14 egorka-P5K audit[32702]: SECCOMP auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="WebKitWebProces" exe="/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKit>
авг 19 18:03:14 egorka-P5K kernel: audit: type=1326 audit(1660914194.306:167): auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="WebKitWebProces" exe="/usr/lib/x86_64-l>
авг 19 18:03:14 egorka-P5K audit[32702]: SECCOMP auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKit>
авг 19 18:03:14 egorka-P5K audit[32702]: SECCOMP auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKit>
авг 19 18:03:14 egorka-P5K audit[32702]: SECCOMP auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKit>
авг 19 18:03:14 egorka-P5K audit[32702]: SECCOMP auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKit>
авг 19 18:03:14 egorka-P5K audit[32702]: SECCOMP auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-linux-gnu/webkit2gtk-4.1/WebKit>
авг 19 18:03:14 egorka-P5K kernel: audit: type=1326 audit(1660914194.490:168): auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-l>
авг 19 18:03:14 egorka-P5K kernel: audit: type=1326 audit(1660914194.490:169): auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-l>
авг 19 18:03:14 egorka-P5K kernel: audit: type=1326 audit(1660914194.490:170): auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-l>
авг 19 18:03:14 egorka-P5K kernel: audit: type=1326 audit(1660914194.490:171): auid=1000 uid=1000 gid=1000 ses=9 subj=? pid=32702 comm="eadedCompositor" exe="/usr/lib/x86_64-l>
авг 19 18:03:14 egorka-P5K xdg-desktop-por[31148]: Realtime error: Could not map pid: Mapping pids is not supported.
авг 19 18:03:14 egorka-P5K audit[32702]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/devices/virtual/dmi/id/chassis_type>
авг 19 18:03:14 egorka-P5K WebKitWebProces[32702]: Could not open /sys/class/dmi/id/chassis_type: Failed to open file “/sys/class/dmi/id/chassis_type”: Permission denied
авг 19 18:03:14 egorka-P5K audit[32702]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/firmware/acpi/pm_profile" pid=32702>
авг 19 18:03:14 egorka-P5K WebKitWebProces[32702]: Could not open /sys/firmware/acpi/pm_profile: Failed to open file “/sys/firmware/acpi/pm_profile”: Permission denied
авг 19 18:03:18 egorka-P5K audit[32685]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/proc/zoneinfo" pid=32685 comm="Pressure>
авг 19 18:03:18 egorka-P5K audit[32685]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/fs/cgroup/user.slice/user-1000.slic>
авг 19 18:03:18 egorka-P5K audit[32685]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/fs/cgroup/user.slice/user-1000.slic>
авг 19 18:03:18 egorka-P5K audit[32685]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/fs/cgroup/user.slice/user-1000.slic>
</pre>

## OS and compositor
<pre>
Ubuntu 22.04 LTS
Compositor - compton
</pre>

I turned to telegram desktop for help, but they sent me here
[https://github.com/telegramdesktop/tdesktop/issues/24953](https://github.com/telegramdesktop/tdesktop/issues/24953)

Telegram Web Apps work great with KDE Plasma, but I don't want to switch to this one

I would like to draw your attention to the error:
<pre>
WebKitNetworkPr[32699]: GDBus.Error:org.freedesktop.portal.Error.NotAllowed: This call is not available inside the sandbox. 
</pre>
It wasn't in KDE Plasma

**Comment by Airblader** [comment]
I see no reason whatsoever to assume that this has something to do with i3. Why would i3 be involved in any DBUS stuff? 

I'll convert this to a discussion because there's nothing to investigate on the window manager side of things. KDE isn't just a window manager.

**Reply by ilya-fedin** [reply]
Window embedding is involved here, not dbus. Content is displayed by embedded window of webkit2gtk. i3 doesn't display the embedded window for @WhatIsThisBreakfast for whatever reason.

**Reply by Airblader** [reply]
I don't know what you mean with "embedded", but the window manager only cares about top level windows. We're not responsible for what happens inside the client's window.

**Reply by ilya-fedin** [reply]
I mean this: https://doc.qt.io/qt-5/qwidget.html#createWindowContainer. The content of top-level window is embedded into other window.

**Reply by Airblader** [reply]
> The container takes over ownership of window. 

The second the window is unmapped as a top level window i3 will not know or care about it anymore.

In any case, OP can easily rule out i3 by testing with some other standalone window manager like awesome or openbox.

**Reply by ilya-fedin** [reply]
He says it works with kwin

> Telegram Web Apps work great with KDE Plasma, but I don't want to switch to this one

**Reply by orestisfl** [reply]
KDE Plasma is not a "standalone window manager"

**Reply by ilya-fedin** [reply]
What if kwin is the only WM handling reparenting (I guess that's what abstraced under Qt's window embedding) right (at least with Qt)?

**Reply by orestisfl** [reply]
> What if kwin is the only WM handling reparenting right

I'd like to see some X11-specific explanation on why this would be the case.

Btw, this seems to be working for me in i3:
![image](https://user-images.githubusercontent.com/5778622/191566069-f3be461a-831f-48ad-b818-7025858b682b.png)

**Reply by ilya-fedin** [reply]
> I'd like to see some X11-specific explanation on why this would be the case.

I don't know how it works under the hood unfortunately, for me as a Qt user this is just a toolkit magic. But for whatever reason there's a known issue (from tdesktop point of view, I don't think anyone reported it to Qt or mutter) that this doesn't work on mutter and its forks (the picture of embedded window just doesn't update while it's presented on the screen, if one minimizes the window and restores it, it changes the content, but won't update it until you repeat that). It also seems xfwm doesn't passthrough scrolling events to the embedded window. So the behavior is really WM-specific here, different WMs show different bugs.

> Btw, this seems to be working for me in i3

Maybe @WhatIsThisBreakfast uses a compositor? I guess it can change the behavior.

**Reply by orestisfl** [reply]
Hmm, interesting with default `picom` it also breaks for me and `picom` outputs:

```
[ 2022-09-21 19:15:46.513 x_create_picture_with_pictfmt_and_pixmap ERROR ] failed to create picture (X error 9 DRAWABLE request 139 minor 4 serial 5088)
[ 2022-09-21 19:15:46.514 paint_one ERROR ] Window 0x0060161c is missing painting data.
[ 2022-09-21 19:15:46.528 x_create_picture_with_pictfmt_and_pixmap ERROR ] failed to create picture (X error 9 DRAWABLE request 139 minor 4 serial 5293)
[ 2022-09-21 19:15:46.530 paint_one ERROR ] Window 0x0060161f is missing painting data.
```

**Reply by orestisfl** [reply]
You can ignore the errors above, they seem related to firefox.

I tried `picom` + `telegram-desktop` + `openbox` and there doesn't seem to be an issue there.

**Comment by laur89** [comment]
Agreed, doesn't sound like this has to do with window-manager. For what it's worth, I'm using Telegram with i3 no-problemo.

**Comment by geeksesi** [comment]
is there any update for it? as i see on https://github.com/telegramdesktop/tdesktop/issues/24953 there should be a issue. but i can't find it.
what happened? i can't also see any webapp.
I'm using manjaro + i3.

![image](https://github.com/i3/i3/assets/28778964/531c0e02-6e7c-44e9-9bb4-4979f2e6d1f0)

```

(telegram-desktop:260718): GLib-GIO-WARNING **: 09:29:33.706: Your application does not implement g_application_activate() and has no handlers connected to the 'activate' signal.  It should do one of these.

(telegram-desktop:260727): GLib-GIO-WARNING **: 09:29:33.906: Your application does not implement g_application_activate() and has no handlers connected to the 'activate' signal.  It should do one of these.

```

**Reply by ilya-fedin** [reply]
i3 developers don't seem to believe they have an issue even though it was confirmed this happens only with i3, not with openbox or anything else...

**Reply by orestisfl** [reply]
> i3 developers don't seem to believe they have an issue even though it was confirmed this happens only with i3, not with openbox or anything else...

Also reproducible in awesomewm

**Comment by waterkip** [comment]
I see this: `apparmor="DENIED"` in your logs:
 
```
авг 19 18:03:14 egorka-P5K audit[32702]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/devices/virtual/dmi/id/chassis_type>
авг 19 18:03:14 egorka-P5K WebKitWebProces[32702]: Could not open /sys/class/dmi/id/chassis_type: Failed to open file “/sys/class/dmi/id/chassis_type”: Permission denied
авг 19 18:03:14 egorka-P5K audit[32702]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/firmware/acpi/pm_profile" pid=32702>
авг 19 18:03:14 egorka-P5K WebKitWebProces[32702]: Could not open /sys/firmware/acpi/pm_profile: Failed to open file “/sys/firmware/acpi/pm_profile”: Permission denied

# And.. 

авг 19 18:03:18 egorka-P5K audit[32685]: AVC apparmor="DENIED" operation="open" profile="snap.telegram-desktop.telegram-desktop" name="/sys/fs/cgroup/user.slice/user-1000.slic>
```
Could that have anything to do with it?

**Reply by ilya-fedin** [reply]
That's just hardware type info, that's unlikely to be critical. I guess anyone using snap have this.

**Comment by ilya-fedin** [comment]
In case anyone needs, I finally found time to report to Qt a similar (or maybe it's actually the same) issue that manifests on mutter. It has a minimal Qt code that leads to the same observable result on mutter. Feel free to try it, if it leads to the same effect on i3, you can vote for it on Qt site, leave a comment that i3 is also affected or maybe try to debug it from i3/picom side.
https://bugreports.qt.io/browse/QTBUG-125724

**Reply by ilya-fedin** [reply]
They repurposed the report for their own purposes so I updated the link

**Comment by LWJerri** [comment]
Hello, just faced with same problem. Solved (idk is this best way) this just running Telegram from command line with `WEBKIT_DISABLE_COMPOSITING_MODE=1` argument.
